File: prime/src/main/java/org/ostelco/prime/ocs/ActivateResponseHolder.java
package org.ostelco.prime.ocs;

import io.grpc.stub.StreamObserver;
import org.ostelco.ocs.api.ActivateResponse;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * Helper class to keep track of
 * {@link io.grpc.stub.StreamObserver<org.ostelco.ocs.api.ActivateResponse>}
 * instance in a threadsafe manner.
 */
final class ActivateResponseHolder {

    private final Lock readLock;

    private final Lock writeLock;

    private StreamObserver<ActivateResponse> activateResponse;

    ActivateResponseHolder() {
        final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();

        this.readLock = readWriteLock.readLock();
        this.writeLock = readWriteLock.writeLock();
    }

    public void setActivateResponse(final StreamObserver<ActivateResponse> ar) {
        writeLock.lock();
        try {
            activateResponse = ar;
        } finally {
            writeLock.unlock();
        }
    }

    public void onNextResponse(final ActivateResponse response) {
        readLock.lock();
        try {
            if (activateResponse != null) {
                activateResponse.onNext(response);
            }
        } finally {
            readLock.unlock();
        }
    }
}

File: ocs/src/main/kotlin/org/ostelco/prime/ocs/ActivateResponseHolder.kt
package org.ostelco.prime.ocs

import io.grpc.stub.StreamObserver
import org.ostelco.ocs.api.ActivateResponse
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantReadWriteLock

/**
 * Helper class to keep track of
 * [<]
 * instance in a threadsafe manner.
 */
internal class ActivateResponseHolder {

    private val readLock: Lock

    private val writeLock: Lock

    private var activateResponse: StreamObserver<ActivateResponse>? = null

    init {
        val readWriteLock = ReentrantReadWriteLock()

        this.readLock = readWriteLock.readLock()
        this.writeLock = readWriteLock.writeLock()
    }

    fun setActivateResponse(ar: StreamObserver<ActivateResponse>) {
        writeLock.lock()
        try {
            activateResponse = ar
        } finally {
            writeLock.unlock()
        }
    }

    fun onNextResponse(response: ActivateResponse) {
        readLock.lock()
        try {
            if (activateResponse != null) {
                activateResponse!!.onNext(response)
            }
        } finally {
            readLock.unlock()
        }
    }
}
